pkg/admission/apiexportendpointslice/apiexportendpointslice_admission.go

```diff
@@ -0,0 +1,266 @@
+			_, err := o.getAPIExport(path, slice.Spec.APIExport.Name)
+			if err != nil {
+				return forbidden
+			}
+		}
+	}
+
+	return nil
+}
+
+// Validate validates the creation of APIExportEndpointSlice resources. It also performs a SubjectAccessReview
+// making sure the user is allowed to use the 'bind' verb with the referenced APIExport.
+func (o *apiExportEndpointSliceAdmission) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) error {
+	clusterName, err := genericapirequest.ClusterNameFrom(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	if a.GetResource().GroupResource() != apisv1alpha1.Resource("apiexportendpointslices") {
+		return nil
+	}
+
+	u, ok := a.GetObject().(*unstructured.Unstructured)
+	if !ok {
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+
+	slice := &apisv1alpha1.APIExportEndpointSlice{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, slice); err != nil {
+		return fmt.Errorf("failed to convert unstructured to APIExportEndpointSlice: %w", err)
+	}
+
+	// Object validation
+	var errs field.ErrorList
+	var oldSlice *apisv1alpha1.APIExportEndpointSlice
+	switch a.GetOperation() {
+	case admission.Create:
+		errs = ValidateAPIExportEndpointSlice(slice)
+		if len(errs) > 0 {
+			return admission.NewForbidden(a, fmt.Errorf("%v", errs))
+		}
+	case admission.Update:
+		u, ok = a.GetOldObject().(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("unexpected type %T", a.GetOldObject())
+		}
+		oldSlice = &apisv1alpha1.APIExportEndpointSlice{}
+		if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, oldSlice); err != nil {
+			return fmt.Errorf("failed to convert unstructured to APIExportEndpointSlice: %w", err)
+		}
+	}
+
+	switch {
+	case a.GetOperation() == admission.Create,
+		// this may be redundant as APIExportEndpointSlice.Spec.APIExport is immutable
+		a.GetOperation() == admission.Update && !reflect.DeepEqual(slice.Spec.APIExport, oldSlice.Spec.APIExport):
+		// unified forbidden error that does not leak workspace existence
+		// only considering "create" as the APIExport reference is immutable
+		forbidden := admission.NewForbidden(a, fmt.Errorf("unable to create APIExportEndpointSlice: no permission to bind to export %s",
+			logicalcluster.NewPath(slice.Spec.APIExport.Path).Join(slice.Spec.APIExport.Name).String()))
+
+		// get cluster name of export
+		var exportClusterName logicalcluster.Name
+		if slice.Spec.APIExport.Path == "" {
+			exportClusterName = clusterName
+		} else {
+			path := logicalcluster.NewPath(slice.Spec.APIExport.Path)
+			export, err := o.getAPIExport(path, slice.Spec.APIExport.Name)
+			if err != nil {
+				return forbidden
+			}
+			exportClusterName = logicalcluster.From(export)
+		}
+
+		// Access check
+		if err := o.checkAPIExportAccess(ctx, a.GetUserInfo(), exportClusterName, slice.Spec.APIExport.Name); err != nil {
```

Comment:
what if an APIExport wasn't set ? I think we should check for it before we end up here

Reply:
It is a mandatory field: https://github.com/kcp-dev/kcp/blob/main/pkg/apis/apis/v1alpha1/types_apiexportendpointslice.go#L55-L60
if we still want to check that I probably need to make it a pointer